pilot/cmd/pilot-agent/status/server.go

```diff
@@ -502,8 +514,56 @@ func negotiateMetricsFormat(contentType string) expfmt.Format {
 	return expfmt.FmtText
 }
 
-func processMetrics(metrics []byte) []byte {
-	return bytes.ReplaceAll(metrics, []byte("\n\n"), []byte("\n"))
+// metricReader used to replace "\n\n" to '\n'
+type metricReader struct {
+	buf *bufio.Reader
+}
+
+func NewMetricReader(reader io.Reader) *metricReader {
+	return &metricReader{buf: bufio.NewReader(reader)}
+}
+
+func (r *metricReader) WriteTo(w io.Writer) (n int64, err error) {
+	var length int
+	var line []byte
+	newLine := []byte{'\n'}
+
+	for ; err == nil; line, _, err = r.buf.ReadLine() {
+		if len(line) == 0 {
+			continue
+		}
+		length, err = w.Write(line)
+		n += int64(length)
+		if err != nil {
+			return
+		}
+		length, err = w.Write(newLine)
+		n += int64(length)
+	}
+	return
+}
+
+func (r *metricReader) Read(p []byte) (n int, err error) {
+	var line []byte
+	for ; err == nil; line, _, err = r.buf.ReadLine() {
+		if len(line) == 0 {
+			continue
+		}
+		if n+len(line)+1 > cap(p) {
+			for range line {
+				_ = r.buf.UnreadByte()
+			}
+			return
+		}
+		copy(p[n:], line)
+		p[n+len(line)] = '\n'
```

Comment:
If we are going to add this custom parsing logic I think we need a LOT of testing and benchmarks to prove this is worth the complexity